lib/date_time/date_time.c

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <date_time.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <device.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <at_cmd.h>
+#include <time.h>
+#include <errno.h>
+#include <string.h>
+#include <net/sntp.h>
+#include <net/socketutils.h>
+#include <sys/timeutil.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(date_time, CONFIG_DATE_TIME_LOG_LEVEL);
+
+#define AT_CMD_MODEM_DATE_TIME			"AT+CCLK?"
+#define AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN	28
+
+/* The magic number 115 corresponds to the year 2015 which is the default
+ * year given by the modem when the cellular network has not pushed time
+ * the modem.
+ */
+#define MODEM_TIME_DEFAULT 115
+
+#define UIO_IP      "129.240.2.6"
+#define GOOGLE_IP   "216.239.35.0"
+#define GOOGLE_IP_2 "216.239.35.4"
+#define GOOGLE_IP_3 "216.239.35.8"
+#define GOOGLE_IP_4 "216.239.35.12"
+
+#define NTP_DEFAULT_PORT "123"
+
+K_SEM_DEFINE(time_fetch_sem, 0, 1);
+
+static struct k_delayed_work time_work;
+
+static struct time_aux {
+	s64_t date_time_utc;
+	int last_date_time_update;
+} time_aux;
+
+static struct sntp_time sntp_time;
+static bool initial_valid_time;
+
+struct ntp_servers {
+	const char *server;
+	struct addrinfo *addr;
+};
+
+struct ntp_servers servers[] = {
+	{.server = UIO_IP},
+	{.server = GOOGLE_IP},
+	{.server = GOOGLE_IP_2},
+	{.server = GOOGLE_IP_3},
+	{.server = GOOGLE_IP_4}
+};
+
+#if defined(CONFIG_DATE_TIME_MODEM)
+static int time_entries_parse(char *datetime_string, int min, int max)
+{
+	char buf[50];
+
+	memcpy(&buf, &datetime_string[min], max - min + 1);
+
+	buf[max - min + 1] = '\0';
+
+	return strtol(buf, NULL, 10);
+}
+
+static int time_modem_get(void)
+{
+	int err;
+	char buf[AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN + 5];
+	struct tm date_time;
+
+	err = at_cmd_write(AT_CMD_MODEM_DATE_TIME, buf, sizeof(buf), NULL);
+	if (err) {
+		LOG_DBG("Could not get cellular network time, error: %d", err);
+		return err;
+	}
+	/* This line zero indexes the buffer at the deisred length
+	 * This ensures clean printing of the modem response.
+	 */
+	buf[AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN] = '\0';
+
+	/* Example of modem time response: +CCLK: "20/02/24,13:44:49+04"
+	 */
+	LOG_DBG("Response from modem: %s", log_strdup(buf));
+
+	date_time.tm_year = time_entries_parse(buf, 8, 9) + 2000 - 1900;
+	date_time.tm_mon  = time_entries_parse(buf, 11, 12) - 1;
+	date_time.tm_mday = time_entries_parse(buf, 14, 15);
+	date_time.tm_hour = time_entries_parse(buf, 17, 18);
+	date_time.tm_min  = time_entries_parse(buf, 20, 21);
+	date_time.tm_sec  = time_entries_parse(buf, 23, 24);
+
```

Comment:
I don't see the need to copy segments of buf into a separate buffer before 'strtol'ing it.
The below will do and is not subject to the wrong indexes used, e.g. if one number could
be bigger than what we assumed it could be here.

```
// Replace '/' ',' and ':' with whitespace for easier parsing by strtol. strtol skips over whitespace.

for(int i = 0; i < AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN; i++) {
  if(buf[i] == '/' || buf[i] == ',' || buf[i] == ':') {
    buf[i] = ' ';
}
}

char* ptr_index = buf;
int base = 10;

date_time.tm_year = strtol(ptr_index, &ptr_index, base) + 2000 - 1900;
date_time.tm_mon = strtol(ptr_index, &ptr_index, base) - 1;


```

Reply:
Thank you! This worked like a charm.